lua/starfall/libs_sh/quaternion.lua

```diff
@@ -1,681 +1,827 @@
+function quat_meta.__unm(q)
+	return wrap({ -q[1], -q[2], -q[3], -q[4] })
 end
 
---- Copies from quaternion and returns a new quaternion
--- @return The copy of the quaternion
-function quat_methods:clone()
-	return wrap({ self[1], self[2], self[3], self[4] })
+--- equivalence metamethod
+-- @param lhs Left side of equation
+-- @param rhs Right side of equation
+-- @return True if both sides are equal
+function quat_meta.__eq(lhs, rhs)
+	return lhs[1] == rhs[1] and lhs[2] == rhs[2] and lhs[3] == rhs[3] and lhs[4] == rhs[4]
 end
 
---- Copies a quaternion to another.
--- @param b The quaternion to copy from.
--- @return nil
-function quat_methods:set(b)
-	self[1] = b[1]
-	self[2] = b[2]
-	self[3] = b[3]
-	self[4] = b[4]
+--- tostring metamethod
+-- @param q Quaternion
+-- @return Quaternion represented as a string
+function quat_meta.__tostring(q)
+	return table.concat(unwrap(q), " ", 1, 4)
 end
 
---- Returns the conj of self
-function quat_methods:conj()
-	return quat_lib.conj(self)
-end
+-------------------------------------
 
-function quat_methods:inv()
-	return quat_lib.inv(self)
+--- Returns components of the quaternion
+-- @return r, i, j, k
+function quat_methods:unpack()
+	return qunpack(self)
 end
 
---- Returns the real component of the quaternion
-function quat_methods:real()
-	return self[1]
+--- Creates a copy of the quaternion
+-- @return Duplicate quaternion
+function quat_methods:clone()
+	return wrap({ qunpack(self) })
 end
 
---- Alias for :real() as r is easier
-function quat_methods:r()
-	return self:real()
+--- Copies components of the second quaternion to the first quaternion. Self-modifies
+-- @param quat Quaternion to copy from
+function quat_methods:set(quat)
+	self[1] = quat[1]
+	self[2] = quat[2]
+	self[3] = quat[3]
+	self[4] = quat[4]
 end
 
+--- Sets R (real) component of the quaternion
+-- @param r Value of the R component
+function quat_methods:setR(r)
+	self[1] = r
+end
 
---- Returns the i component of the quaternion
-function quat_methods:i()
-	return self[2]
+--- Sets I component of the quaternion
+-- @param i Value of the I component
+function quat_methods:setI(i)
+	self[2] = i
 end
 
---- Returns the j component of the quaternion
-function quat_methods:j()
-	return self[3]
+--- Sets J component of the quaternion
+-- @param j Value of the J component
+function quat_methods:setJ(j)
+	self[3] = j
 end
 
---- Returns the k component of the quaternion
-function quat_methods:k()
-	return self[4]
+--- Sets K component of the quaternion
+-- @param k Value of the K component
+function quat_methods:setK(k)
+	self[4] = k
 end
 
---[[****************************************************************************]]
+-------------------------------------
 
---- Raises Euler's constant e to the power <q>
-function quat_lib.exp(q)
-	return qexp(q)
+--- Raises Euler's constant e to the quaternion's power
+-- @return Constant e raised to the quaternion
+function quat_methods:getExp()
+	return wrap(quatExp(unwrap(self)))
```

Comment:
Not sure if I can avoid using unwrap here and in similar functions like `q.normalize`
So please also check that.

Reply:
You don't need an unwrap. And just add wrap({data}) inside each of your local functions that create a new quaternion. Then this would just be `return quatExp(self)`

Reply:
In-fact I would just rename 'wrap' to newquat or whatever you had before.

Reply:
If I did that then some quaternions will get wrapped multiple times in one call.
Take `quatMul` for example, it is sometimes called like that: `quatMul(quatMul(a, b), c)` and often tables returned by helper functions are reconstructed into something completely new that would have to be wrapped again anyway.
Do you think that will be ok?

Reply:
You can create functions that self modify and functions that return new quaternions. The self modify ones will avoid creating new tables and such.

`quatMul(getQuatMul(a, b), c)`

where `getQuatMul` is
```
function getQuatMul(a, b)
    a = a:clone()
    quatMul(a,b)
    return a
end
```

Reply:
I see now you were using `unwrap` as your 'clone'.

Reply:
I kinda want to kill you Sparky.
![violent_death_by_gun](https://media1.tenor.com/images/76e17ebaecb2f48071a1dec366a91c68/tenor.gif)

I don't know why I listened to your idea about self-modifying everything. Had to basically rewrite most of the code.
I probably overcomplicated stuff, but I don't want to hear a no no, it works and it uses as little tables as possible.
Side effect is that the code is completely retarded in some places, but u gotta do what u gotta do ¯\\\_(ツ)\_/¯

Reply:
Lols, welcome to mutability. This is why const is important in c++